jax/_src/array.py

```diff
@@ -680,54 +680,85 @@ def make_array_from_callback(
-  addressable and belong to the current process.
+  To use this function, every device in the global mesh must have an addressable array that
+  is already on on-device, typically either via ``jax.device_put`` or generated directly onto the device.
+
+  The ``jax.Array``\s forming the input argument ``arrays`` must be local to the calling
+  process.
 
   Args:
-    shape : Shape of the ``jax.Array``.
-    sharding: A ``Sharding`` instance which describes how the ``jax.Array`` is
-      laid out across devices.
-    arrays: Sequence of ``jax.Array``\s that are on a single device.
+    shape : Shape of the output ``jax.Array``. This is redundant with ``sharding`` and ``arrays`` and serves as a
+      double check.
+    sharding: A ``Sharding`` instance which describes how the output ``jax.Array`` is
+      shaped and replicated.
+    arrays: Sequence of ``jax.Array``\s that are each on a single device for this process. ``len(arrays)``
+      must equal ``len(jax.local_devices())`` and the shape of each array must be the same. For multiprocess code,
+      the each process will call with a different ``arrays`` argument that corresponds to that processes' data.
 
   Returns:
-    A ``jax.Array`` from a sequence of ``jax.Array``\s on a single device.
+    A global ``jax.Array``, laid out based on ``sharding`` of shape equal to ``shape``, and with contents per device
+      matching ``arrays``.
 
-  Example:
+  Examples:
+
+    In this single process example, we use ``make_array_from_single_device_arrays`` to create an
+    a global array
 
     >>> import math
     >>> from jax.sharding import Mesh
     >>> from jax.sharding import PartitionSpec as P
     >>> import numpy as np
     ...
+    >>> mesh_rows = 2
+    >>> mesh_cols =  jax.device_count()//2
+    ...
     >>> global_shape = (8, 8)
-    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(2, 4), ('x', 'y'))
+    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(mesh_rows, mesh_cols), ('x', 'y'))
     >>> sharding = jax.sharding.NamedSharding(global_mesh, P('x', 'y'))
     >>> inp_data = np.arange(math.prod(global_shape)).reshape(global_shape)
     ...
     >>> arrays = [
-    ...     jax.device_put(inp_data[index], d)
-    ...     for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
+    ...    jax.device_put(inp_data[index], d)
+    ...        for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
     ...
     >>> arr = jax.make_array_from_single_device_arrays(global_shape, sharding, arrays)
-    >>> arr.addressable_data(0).shape
-    (4, 2)
-
-    In multi-process case, if the input is process local and data parallel
-    i.e. each process receives a different part of the data, then you can use
-    `make_array_from_single_device_arrays` to create a global jax.Array
+    >>> print(f"{arr.addressable_data(0).shape=}")
+    # (4,4) if jax.device_count()==4.
```

Comment:
The whole point of doctest is to not have comments and this code should be tested. So please, remove the comment and make it testable.

Reply:
You don't write it for both? This is just an example. Users can adapt it to their use case.

This doctest is going to run on a CPU with 8 devices. So you can just assume that and test via this: https://jax.readthedocs.io/en/latest/developer.html#doctests

Reply:
I disagree with this. As a user I assumed there had been bitrot when I couldn't run this on my v4-8. Can we figure out how to write code that a user can run?

Reply:
Well, then run the example with 4 devices. You can do `jax.devices()[:4]`. I don't understand why you are assuming things to only run on v4. `v3` still exists in the world :)

Reply:
Rafi TODO: add the assertions to Python!